### PR TITLE
enable all warnings and add developer mode

### PR DESCRIPTION
**Pull request by pbalcer**
For now, the developer mode only turns warnings into errors, but over time it will also encompass other various checks that we don't want to happen in
normal builds.

I also did a small cleanup of the main cmake.
The UR_ENABLE_TESTS option is now BUILD_TESTS.

### CODE REVIEW

**Comment by pbalcer** [comment]
I also wanted to enable asan and ubsan, but that's going to be its own PR since it's more work than I expected.

**Review by veselypeta — APPROVED** [review]
LGTM

**Comment by DamianDuy** [comment]
From what I see on Windows check on github, there's a lot of linker warnings in logs. Shouldn't this fail since the -WX option is on?

File: .github/workflows/cmake.yml [hunk 1]
```diff
@@ -26,7 +22,7 @@ jobs:
       run: pip install -r third_party/requirements.txt
 
     - name: Configure CMake
-      run: cmake -B ${{github.workspace}}/build -DCMAKE_BUILD_TYPE=${{env.BUILD_TYPE}} -DUR_ENABLE_TESTS=ON
+      run: cmake -B ${{github.workspace}}/build -DDEVELOPER_MODE=ON -DCMAKE_BUILD_TYPE=${{env.BUILD_TYPE}} -DBUILD_TESTS=ON
```
Comment (PatKamin) [inline .github/workflows/cmake.yml:25]: Add DEVELOPER_MODE to the list of cmake options in the README (perhaps with info that it enables Werror compiler flag?)

Comment (pbalcer) [reply .github/workflows/cmake.yml:25]: done

**Pull request merged**